[scroll-animations] account for start and end delays for progress-based animations

graouts opened the pull request:
#### 086d3e1d9aebb93b28201fdc219a1bafde5c3028
<pre>
[scroll-animations] account for start and end delays for progress-based animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=283790">https://bugs.webkit.org/show_bug.cgi?id=283790</a>
<a href="https://rdar.apple.com/140650005">rdar://140650005</a>

Reviewed by Dean Jackson.

So far, we have always assumed start and end delays for progress-based animations are irrelevant.
However, the procedure to normalize specified timing [0], which we partially implemented under
`AnimationEffect::normalizeSpecifiedTiming()`, calls out how to implement this. This incomplete
implementation causes a number of WPT failures.

To correctly normalize specified timing, we need to add the notion of a &quot;specified start delay&quot;,
&quot;specified end delay&quot; and &quot;specified iteration duration&quot;. Those are the values that come through
the JavaScript API or the relevant CSS properties. Normalizing the specified timing relies on those
properties as well as the animation playback rate, the timeline duration and the iteration count.

To correctly normalize the specified timing when required, we now add a new `m_timingDidMutate`
instance variable to `AnimationEffect` to track when one of those properties has been changed.
Then, we call `updateComputedTimingPropertiesIfNeeded()` when any of the timing properties that
depend on those is queried. This means that some methods that used to be `const` no longer are.

This in itself is an improvement because it allows us to remove the poorly-named public method
`AnimationEffect::updateStaticTimingProperties()` which was called outside of `AnimationEffect`
when other parts of the code assumed that timing may have been changed and computed properties
needed an update. Additionally, we remove the `AnimationEffect::normalizeSpecifiedTiming()`
method since this should be within the purview of the supporting `AnimationEffectTiming` object.

Now, `AnimationEffectTiming::updateComputedProperties()` deals with the normalization step as a
prelude to computing the &quot;active duration&quot; and the &quot;end time&quot;. Correctly dealing with delays for
progress-based animations also allows us to remove some special casing for percentage timing values
throughout the rest of `AnimationEffectTiming`.

This yields significant WPT progress with 57 new PASS results, one unskipped reftest on macOS and one
flaky test now proving to be stable on all platforms.

[0] <a href="https://drafts.csswg.org/web-animations-2/#normalize-specified-timing">https://drafts.csswg.org/web-animations-2/#normalize-specified-timing</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-ignored.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/progress-based-animation-animation-longhand-properties.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/effect-updateTiming-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/intrinsic-iteration-duration.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-animation-effect-fill-modes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-animation-effect-phases.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/zero-intrinsic-iteration-duration.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animation/scroll-timeline-overallProgress.tentative-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getBasicTiming):
(WebCore::AnimationEffect::getBindingsComputedTiming):
(WebCore::AnimationEffect::getComputedTiming):
(WebCore::AnimationEffect::updateTiming):
(WebCore::AnimationEffect::setIterations):
(WebCore::AnimationEffect::delay):
(WebCore::AnimationEffect::setDelay):
(WebCore::AnimationEffect::endDelay):
(WebCore::AnimationEffect::setEndDelay):
(WebCore::AnimationEffect::iterationDuration):
(WebCore::AnimationEffect::setIterationDuration):
(WebCore::AnimationEffect::activeDuration):
(WebCore::AnimationEffect::endTime):
(WebCore::AnimationEffect::timeToNextTick):
(WebCore::AnimationEffect::animationTimelineDidChange):
(WebCore::AnimationEffect::animationPlaybackRateDidChange):
(WebCore::AnimationEffect::updateComputedTimingPropertiesIfNeeded):
(WebCore::durationAPIValue): Deleted.
(WebCore::AnimationEffect::getBasicTiming const): Deleted.
(WebCore::AnimationEffect::getBindingsComputedTiming const): Deleted.
(WebCore::AnimationEffect::getComputedTiming const): Deleted.
(WebCore::AnimationEffect::normalizeSpecifiedTiming): Deleted.
(WebCore::AnimationEffect::updateStaticTimingProperties): Deleted.
(WebCore::AnimationEffect::timeToNextTick const): Deleted.
* Source/WebCore/animation/AnimationEffect.h:
(WebCore::AnimationEffect::specifiedDelay const):
(WebCore::AnimationEffect::specifiedEndDelay const):
(WebCore::AnimationEffect::specifiedIterationDuration const):
(WebCore::AnimationEffect::delay const): Deleted.
(WebCore::AnimationEffect::endDelay const): Deleted.
(WebCore::AnimationEffect::iterationDuration const): Deleted.
(WebCore::AnimationEffect::activeDuration const): Deleted.
(WebCore::AnimationEffect::endTime const): Deleted.
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::updateComputedProperties):
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::setTimingProperties):
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::copyPropertiesFromSource):
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer):
(WebCore::KeyframeEffect::timeToNextTick):
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer const): Deleted.
(WebCore::KeyframeEffect::timeToNextTick const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setPlaybackRate):
(WebCore::WebAnimation::applyPendingPlaybackRate):
(WebCore::WebAnimation::playState const):
(WebCore::WebAnimation::updateFinishedState):
(WebCore::WebAnimation::overallProgress const):
* Source/WebCore/animation/WebAnimationTime.cpp:
(WebCore::WebAnimationTime::isNaN const):
(WebCore::WebAnimationTime::matchingInfinity const):
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/287181@main">https://commits.webkit.org/287181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3026d4174fd3e6b9f07899b22e74e228cd295560

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78709 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32091 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83369 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29971 "Failed to checkout and rebase branch from PR 37219") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66905 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6034 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/29971 "Failed to checkout and rebase branch from PR 37219") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81776 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/51673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/70825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/41951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/25674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28311 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/26055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6074 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6236 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/69123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/13162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/11744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12144 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6019 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/11934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9441 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7793 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/animation/AnimationEffect.cpp:
```diff
@@ -110,30 +93,46 @@ AnimationEffectTiming::ResolutionData AnimationEffect::resolutionData(std::optio
+    // is the string auto, this attribute will return the current calculated value of the intrinsic iteration
+    // duration, which may be a expressed as a double representing the duration in milliseconds or a percentage
+    // when the effect is associated with a progress-based timeline.
+    auto computedDuration = [&]() -> DoubleOrCSSNumericValueOrString {
```
What's the point of making this a function? You only call it once with no arguments, so call it here?

grorg replied:
I mean, making the variable a function, rather than just the value.

graouts replied:
No point, I'll make the suggested change.

graouts committed:
[scroll-animations] account for start and end delays for progress-based animations
https://bugs.webkit.org/show_bug.cgi?id=283790
rdar://140650005

Reviewed by Dean Jackson.

So far, we have always assumed start and end delays for progress-based animations are irrelevant.
However, the procedure to normalize specified timing [0], which we partially implemented under
`AnimationEffect::normalizeSpecifiedTiming()`, calls out how to implement this. This incomplete
implementation causes a number of WPT failures.

To correctly normalize specified timing, we need to add the notion of a "specified start delay",
"specified end delay" and "specified iteration duration". Those are the values that come through
the JavaScript API or the relevant CSS properties. Normalizing the specified timing relies on those
properties as well as the animation playback rate, the timeline duration and the iteration count.

To correctly normalize the specified timing when required, we now add a new `m_timingDidMutate`
instance variable to `AnimationEffect` to track when one of those properties has been changed.
Then, we call `updateComputedTimingPropertiesIfNeeded()` when any of the timing properties that
depend on those is queried. This means that some methods that used to be `const` no longer are.

This in itself is an improvement because it allows us to remove the poorly-named public method
`AnimationEffect::updateStaticTimingProperties()` which was called outside of `AnimationEffect`
when other parts of the code assumed that timing may have been changed and computed properties
needed an update. Additionally, we remove the `AnimationEffect::normalizeSpecifiedTiming()`
method since this should be within the purview of the supporting `AnimationEffectTiming` object.

Now, `AnimationEffectTiming::updateComputedProperties()` deals with the normalization step as a
prelude to computing the "active duration" and the "end time". Correctly dealing with delays for
progress-based animations also allows us to remove some special casing for percentage timing values
throughout the rest of `AnimationEffectTiming`.

This yields significant WPT progress with 57 new PASS results, one unskipped reftest on macOS and one
flaky test now proving to be stable on all platforms.

[0] https://drafts.csswg.org/web-animations-2/#normalize-specified-timing

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-ignored.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/progress-based-animation-animation-longhand-properties.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/effect-updateTiming-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/intrinsic-iteration-duration.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-animation-effect-fill-modes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/scroll-animation-effect-phases.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/zero-intrinsic-iteration-duration.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/interfaces/Animation/scroll-timeline-overallProgress.tentative-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::getBindingsTiming const):
(WebCore::AnimationEffect::getBasicTiming):
(WebCore::AnimationEffect::getBindingsComputedTiming):
(WebCore::AnimationEffect::getComputedTiming):
(WebCore::AnimationEffect::updateTiming):
(WebCore::AnimationEffect::setIterations):
(WebCore::AnimationEffect::delay):
(WebCore::AnimationEffect::setDelay):
(WebCore::AnimationEffect::endDelay):
(WebCore::AnimationEffect::setEndDelay):
(WebCore::AnimationEffect::iterationDuration):
(WebCore::AnimationEffect::setIterationDuration):
(WebCore::AnimationEffect::activeDuration):
(WebCore::AnimationEffect::endTime):
(WebCore::AnimationEffect::timeToNextTick):
(WebCore::AnimationEffect::animationTimelineDidChange):
(WebCore::AnimationEffect::animationPlaybackRateDidChange):
(WebCore::AnimationEffect::updateComputedTimingPropertiesIfNeeded):
(WebCore::durationAPIValue): Deleted.
(WebCore::AnimationEffect::getBasicTiming const): Deleted.
(WebCore::AnimationEffect::getBindingsComputedTiming const): Deleted.
(WebCore::AnimationEffect::getComputedTiming const): Deleted.
(WebCore::AnimationEffect::normalizeSpecifiedTiming): Deleted.
(WebCore::AnimationEffect::updateStaticTimingProperties): Deleted.
(WebCore::AnimationEffect::timeToNextTick const): Deleted.
* Source/WebCore/animation/AnimationEffect.h:
(WebCore::AnimationEffect::specifiedDelay const):
(WebCore::AnimationEffect::specifiedEndDelay const):
(WebCore::AnimationEffect::specifiedIterationDuration const):
(WebCore::AnimationEffect::delay const): Deleted.
(WebCore::AnimationEffect::endDelay const): Deleted.
(WebCore::AnimationEffect::iterationDuration const): Deleted.
(WebCore::AnimationEffect::activeDuration const): Deleted.
(WebCore::AnimationEffect::endTime const): Deleted.
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::updateComputedProperties):
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::setTimingProperties):
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::copyPropertiesFromSource):
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer):
(WebCore::KeyframeEffect::timeToNextTick):
(WebCore::KeyframeEffect::backingAnimationForCompositedRenderer const): Deleted.
(WebCore::KeyframeEffect::timeToNextTick const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setPlaybackRate):
(WebCore::WebAnimation::applyPendingPlaybackRate):
(WebCore::WebAnimation::playState const):
(WebCore::WebAnimation::updateFinishedState):
(WebCore::WebAnimation::overallProgress const):
* Source/WebCore/animation/WebAnimationTime.cpp:
(WebCore::WebAnimationTime::isNaN const):
(WebCore::WebAnimationTime::matchingInfinity const):
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForEffect):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/287181@main
